[mlir][transform] Add a scoped transform

aniragil opened the pull request:
Add scf.scope modeling a code block and transform.as_scope which applies
transformations over it. The scf.scope op is similar to scf.execute_region
except it's isolated-from-above, which allows applying transforms to a
limited region without resorting to function outlining.

The transform.as_scope op creates a temporary scf.scope containing
clones of its operands and applies the transformations within its body
to that scope. If successful, the scope's results replace all uses of
any original operation outside the scope, the original operations are
erased and the scope is inlined. On failure, the scf.scope is erased.

aniragil committed:
[mlir][transform] Add transform.as_scope op

The transform.as_scope op creates a temporary scf.scope containing
clones of its operands and applies the transformations within its body
to that scope. If successful, the scope's results replace all uses of
any original operation outside the scope, the original operations are
erased and the scope is inlined. On failure, the scf.scope is erased.

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+    scope = dyn_cast<scf::ScopeOp>(*executeRegionBody.getOperations().begin());
+    scope->moveAfter(executeRegion);
+    rewriter.eraseOp(executeRegion);
+#else
```
The (simpler) code for retrieving the (potentially replaced) scf.scope after applying the nested transforms in the #else clause fails for transform.alternatives whose getEffects() consumes the handle it is given (not sure why, as it seems to expect the clones to exist after applying the alternative, @ftynse could you explain?). I resorted to wrapping the scf.scope in a scf.execute_region, but there must be a more elegant way to do this.

ftynse replied:
My hypothesis is the following. The `auto newPayloadOps = state.getPayloadOps(body.front().getArgument(0));` line below gets the payload ops associated with the block argument handle by `mapBlockArguments` above, that is, `scope`. The nested transforms ops may or may not update that association. Specifically, if the scope is rewritten/deleted, the `Operation *` contained in this association will be dangling (but may happen to point to another operation, including the scope that was created as a replacement, by accident of the allocator reusing the pointer address). Hopefully, some of the sanity checks in place will complain that access to a consumed handle is requested (if they don't, let me know and we'll try to ensure they do). The workaround above gets the scope operation directly from the surrounding `execute_region`, which is never rewritten, rather than from the handle association, and is therefore immune to such changes.

The solution to this is to `transform.yield` the (updated) scope handle from the `transform.as_scope` and use the payload associated with the terminator operand to get the most recent scope. Some more thinking is needed to make this interop with `transform.alternative` as that op doesn't return the scope, not sure whether it should return the updated scope or mark the handle as consuming only nested payloads but not the handle itself. I don't think we have a mechanism for the latter, but the need for it was floated around.

aniragil replied:
keeping the wrapping op solution for now, as this seems to require a broader discussion.

ftynse commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+      return DiagnosedSilenceableFailure::definiteFailure();
+    }
+    scopeBody = &scope.getBody();
+    if (!scopeBody->hasOneBlock()) {
+      LLVM_DEBUG(DBGS() << "multiple blocks in scope post transformation\n");
+      return DiagnosedSilenceableFailure::definiteFailure();
+    }
```
Nit: I think this is part of the verifier.

aniragil replied:
Removed

ftynse commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+  }
+
+  rewriter.eraseOp(scope);
+  while (!opsPayload.empty()) {
+    for (auto payload : llvm::make_early_inc_range(opsPayload)) {
+      if (payload->getUsers().empty()) {
+        rewriter.eraseOp(payload);
+        opsPayload.remove(payload);
+      }
+    }
+  }
```
I don't fully understand this. Does this intend to erase original payloads in the right order based on their use-def chains? Can there be a situation where the original payload still has a live user that wasn't updated leading to an infinite loop here? (I suspect there cannot because the value of the payload would be live-out and would have been replaced above, correct?).

aniragil replied:
Right. Added a comment to clarify and an assertion to make sure all users are themselves original payload ops.

aniragil committed:
Update mlir/include/mlir/Dialect/Transform/IR/TransformOps.td

Co-authored-by: Oleksandr "Alex" Zinenko <ftynse@gmail.com>

aniragil committed:
Update mlir/include/mlir/Dialect/Transform/IR/TransformOps.td

Co-authored-by: Oleksandr "Alex" Zinenko <ftynse@gmail.com>

aniragil committed:
Address review comments, add test case

- Fixed and add documentation
- Fixed coding style
- Added test case of an as_scope generated over a single op
- Use createBlock() taking args
- Document and assert forward-progress condition for original payload
  ops: any remaining user after replacing their users with the
  transformed live-outs must be another original payload op.
- Limit user replacement of live-outs to users that do not belong to
  the original payload ops. This is cleaner, even though these ops will
  immediately be deleted.
- Remove hasOneBlock() check already done during verification
- Remove block argument producesHandle code

aniragil commented on mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:
```diff
@@ -931,6 +931,28 @@ def ReduceReturnOp :
+     SingleBlockImplicitTerminator<"scf::YieldOp">]> {
+  let summary = "isolated code scope";
+  let description = [{
+    This op models an executed-once region, similar to `execute_region`
+    except it's isolated from above, which facilitates local scoping of
+    code without resorting to function outlining.
```
Would be great to unify them, but can we make the isolated-from-above trait behave dynamically based on an attribute?

joker-eph replied:
I would be concerned about the IsolatedFromAbove property of an operation not begin a static one.

aniragil commented on mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:
````diff
@@ -434,6 +434,72 @@ def ApplyRegisteredPassOp : TransformDialectOp<"apply_registered_pass",
+    transformation, either for functional reasons, e.g.
+
+    ```mlir
+    transform.as_scope %what before %where : (!transform.any_op, !transform.any_op) -> (), !transform.any_op {
````
Right, fixed

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+// AsScopeOp
+//===----------------------------------------------------------------------===//
+
+static std::optional<std::pair<Operation *, Operation *>>
```
Done

joker-eph commented on mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:
```diff
@@ -931,6 +931,28 @@ def ReduceReturnOp :
+  let description = [{
+    This op models an executed-once region, similar to `execute_region`
+    except it's isolated from above, which facilitates local scoping of
+    code without resorting to function outlining.
```
I don't understand the description, and this the motivation for adding this operation right now, can you elaborate?

Also please split the Pr in two: introduce the op to SCF is orthogonal to the Transform dialect.

aniragil replied:
Sure, some background:
This patch is a draft following a C4ML'24 discussion regarding the proposed transform, which facilitates trying transformation sequences on a limited scope without resorting to function outlining or cloning large portions of the IR using transform.alternatives. This transform requires an isolated-from-above op which can temporarily encapsulate the to-be-transformed part of the code.
The scope op was put in the scf dialect mainly since the transform dialect doesn't allow adding non-transform ops (perhaps it should?). However, a EuroLLVM'24 [round table](https://discourse.llvm.org/t/notes-from-the-mlir-upstream-round-table-eurollvm-2024/78374) brought up potential other use cases for such an op, so I'll split it into its own patch to keep the discussions separate.

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+      earliest = op;
+    else if (latest->isBeforeInBlock(op))
+      latest = op;
+    else
+      ;
```
Modified

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+
+  Operation *wherePayload = nullptr;
+  Value whereOp = getWhereOp();
+  auto whereAttr = getWhere();
```
Done

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+  RelativeLocation where;
+  if (whereOp) {
+    auto wherePayloadOps = state.getPayloadOps(whereOp);
+    if (std::distance(wherePayloadOps.begin(), wherePayloadOps.end()) != 1) {
```
Done

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+  if (whereOp) {
+    auto wherePayloadOps = state.getPayloadOps(whereOp);
+    if (std::distance(wherePayloadOps.begin(), wherePayloadOps.end()) != 1) {
+      auto diag = emitDefiniteFailure()
```
Done

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+  auto scope = rewriter.create<scf::ScopeOp>(wherePayload->getLoc(),
+                                             resultTypes, liveIns);
+  Region *scopeBody = &scope.getBody();
+  // TODO: Move into builder.
```
Removed

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+        break;
+      }
+
+      if (::mlir::failed(result.silence()))
```
Done

aniragil commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+  onlyReadsHandle(getWhereOp(), effects);
+  consumesHandle(getWhat(), effects);
+  producesHandle(getResults(), effects);
+  Region &region = getRegion();
+  if (!region.empty())
+    producesHandle(region.front().getArguments(), effects);
```
Actually copied this from AlternativesOp::getEffects(), but code seems to work find without it so I'll remove it.

aniragil committed:
Remove commented out code

As it is currently unclear how to retrieve the potentially replaced scope
in a robust manner, this option is removed in favor of the current
implementation of wrapping the scope in a scf.execute_region op.

aniragil committed:
[mlir][scf] Add a scope op to the scf dialect

Add to the scf dialect an operation modeling an isolated-from-above
single basic block that is executed once. It provides a localized,
hierarchical alternative to outlining code into a function/call pair.

aniragil commented:
To follow up on some EuroLLVM'24 discussions over this transform: @joker-eph made an interesting suggestion for the transform to use a "floating func.func" as the isolated-from-above payload op instead of the scf.scope op, i.e. outline code into a function not inserted to any module. I like this idea, and @ftynse and @matthias-springer agreed this should work with transforms, but won't this solution prevent the transform from working on operations that use symbols?

ftynse commented:
My concern with detached functions is precisely the fact that they are detached. This sounds quite surprising for logging/debugging because they will not be easily visible when dumping, e.g., the top-level payload module. So I would prefer an inlined isolated-from-above operation instead that is visible in the IR. Furthermore, if we need to send values to or from such a detached function, we would need to keep a call to that function in the IR it was outlined from, which would render that invalid (unless we have a declaration, but at that point we can just as well have the definition there)...
